Adding service component calls to methods in RecordCommunitiesService

monotasker opened the pull request:
:heart: Thank you for your contribution!

### Description

The unique methods in RecordCommunitiesService lack the call to service component methods that are usual in record service operations. This PR 
- adds those calls in the `add`, `remove`, `set_default`, and `bulk_add` methods 
- and adds a `components` property in the service config that reads components from the `RDM_RECORD_COMMUNITIES_SERVICE_COMPONENTS` variable if present (defaults to [])

The placement of the component calls in each method is intended to allow maximum customization of the service call's input data before the data is actually applied to change the records.

Note that in `bulk_add` I've wrapped the `set_default` argument value in a dictionary (making it mutable) so that it can be updated inside the components. 

### Checklist

Ticks in all boxes and 🟢 on all GitHub actions status checks are required to merge:

- [X ] I'm aware of the [code of conduct](https://inveniordm.docs.cern.ch/contribute/code-of-conduct/).
- [X] I've created [logical separate commits](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commits) and followed the [commit message format](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#commit-message).
- [X] I've added relevant test cases.
- [X] I've added relevant documentation.
- [X] I've marked [translation strings](https://inveniordm.docs.cern.ch/develop/howtos/i18n/).
- [X] I've identified the [copyright holder(s)](https://inveniordm.docs.cern.ch/contribute/copyright-policy/) and updated copyright headers for touched files (>15 lines contributions).
- [X] I've NOT included third-party code (copy/pasted source code or new dependencies).
    * If you have added [third-party code (copy/pasted or new dependencies)](https://inveniordm.docs.cern.ch/develop/best-practices/commits/#third-party-codedependencies), please reach out to an [architect on Discord](https://discord.gg/8qatqBC).

**Frontend**

- [X] I've followed the [CSS/JS](https://inveniordm.docs.cern.ch/develop/best-practices/css-js/) and [React](https://inveniordm.docs.cern.ch/develop/best-practices/react/) guidelines.
- [X] I've followed the [web accessibility](https://inveniordm.docs.cern.ch/develop/best-practices/accessibility/) guidelines.
- [X] I've followed the [user interface](https://inveniordm.docs.cern.ch/develop/best-practices/ui/) guidelines.


**Reminder**

By using GitHub, you have already agreed to the [GitHub’s Terms of Service](https://help.github.com/articles/github-terms-of-service/#6-contributions-under-repository-license) including that:

1. You license your contribution under the same terms as the current repository’s license.
2. You agree that you have the right to license your contribution under the current repository’s license.

zzacharo commented on invenio_rdm_records/services/communities/service.py:
```diff
@@ -149,6 +149,11 @@ def add(self, identity, id_, data, uow):
 
+        # Run components
+        for component in self.components:
+            if hasattr(component, "add"):
```
I would align it with the permission so that is more readable in the instance's custom component you implement
```suggestion
            if hasattr(component, "add_community"):
```
What do you think?

zzacharo replied:
discussion: Also, should we add this hook before the `community.add(...)` call or after @slint ?

zzacharo commented on invenio_rdm_records/services/communities/service.py:
```diff
@@ -241,6 +246,19 @@ def remove(self, identity, id_, data, uow):
+
+        # Run components
+        for component in self.components:
+            if hasattr(component, "remove"):
```
I would move this to `_remove` after the permission checks and rename to `remove_community`. Also, the question on where to add the component hook is valid here too.

monotasker commented:
Alright, I've updated the PR and squashed the commits together (with a proper commit message). Sorry for the messy commits earlier, and thanks for your patience.

I renamed the methods as suggested and moved the component calls for `remove` into `_remove`. I have *kept* each of the component calls *before* the data in each case is actually applied to the record. (E.g., before record.parent.communities.add is called in the `add` method.) This is to allow modification of the data before it is applied to the record. I don't see any disadvantages of this placement, and it has the added benefit of providing the component method with the initial state of the record.

tmorrell commented:
The stale tag is automated and doesn't mean anything...beyond we've forgotten about the PR. I'll try to remind folks about this one.

monotasker commented:
Okay @fenekku I made the changes you suggested and updated the single commit (additional updates listed in the new commit message)

tmorrell commented:
Trying to see if there is any last feedback on this...but seeing no concerns we should probably merge this soon.
